Add Akumuli database support

Lazin opened the pull request.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/timescale/tsbs?pullRequest=73) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/timescale/tsbs?pullRequest=73) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/timescale/tsbs?pullRequest=73) it.</sub>

RobAtticus commented:
Hi @Lazin , I see you've updated this (though it does need a few more conflicts fixed). I will try to look into this this week so we can help move it forward, now that CrateDB is merged in.

Lazin commented:
Looking forward to it! I fixed merge conflicts but it looks like something is broken on CI now. I'll investigate this tomorrow.

Lazin commented:
Hi @RobAtticus , 
I updated and tested PR. It can be reviewed now.

blagojts commented:
Hello @Lazin 
The akumuli_test.go file contains some errors (tag values can be any type now, no need to have strings only, and the Point struct accepts an interface{} now)

Lazin commented:
Hi @blagojts 
I updated akumuli_test.go and merged with master.

blagojts commented on cmd/tsbs_load_akumuli/creator.go:
```diff
@@ -0,0 +1,32 @@
+package main
```
If I understood correctly, Akumuli does not have the concept of databases and schemas like PostgreSQL or InfluxDB has. And the db creator is not needed to create a database. But the Creator also handles cleaning of old/stale data (check out tsbs_load_timescaledb/creator.go). I would also change it so that the `DBExists` method always returns `true` so the `RemoveOldDb` gets called by tsbs (where you can put the cleanup code). The calling of RemoveOldDb is controlled by the `do-create-db` flag, which by default is true, but if you want to keep the old data you can set it to false

Lazin replied:
Yes, it doesn't have a notion of database/table/schema. And it's not possible to clean up using the database API. You have to run command in terminal to crate or delete the database.

Lazin commented on cmd/tsbs_generate_data/serialize/akumuli.go:
```diff
@@ -0,0 +1,133 @@
+// AKUMULI RESP protocol.  Serialized adds extra data to guide data loader.
+//
+func (s *AkumuliSerializer) Serialize(p *Point, w io.Writer) (err error) {
+	if s.book == nil {
```
Done

Lazin commented on cmd/tsbs_generate_data/serialize/akumuli_test.go:
```diff
@@ -0,0 +1,115 @@
+
+	for _, c := range cases {
+		actualCnt := strings.Count(got, c.expValue)
+		if actualCnt != c.expCount {
```
It's checked on line 73. If one of the test samples is not present in the output required number of times it will fail.

Lazin commented on cmd/tsbs_generate_queries/databases/akumuli/devops.go:
```diff
@@ -0,0 +1,294 @@
+package akumuli
```
It's not possible to implement using Akumuli query API. Date/time truncation is not supported.

Lazin commented:
@blagojts thanks, I appreciated the feedback!
To run the test you should use recent docker image. 0.7.45 is way too old. The latest tag 0.8.69-skylake should work fine.

RobAtticus commented:
Hey @Lazin , we're seeing failures in CI:
```
# github.com/timescale/tsbs/cmd/tsbs_load_akumuli
cmd/tsbs_load_akumuli/main.go:33:34: not enough arguments in call to load.GetBenchmarkRunner
	have ()
	want (load.BenchmarkRunnerConfig)
```
and
```
# github.com/timescale/tsbs/cmd/tsbs_run_queries_akumuli
cmd/tsbs_run_queries_akumuli/main.go:26:35: not enough arguments in call to query.NewBenchmarkRunner
	have ()
	want (query.BenchmarkRunnerConfig)
```

Lazin commented:
Hi @RobAtticus,
I fixed this errors and ran all benchmark apps to check if anything else was broken. It all worked for me. Also, updated the doc since the command line interface changed slightly.

RobAtticus commented on cmd/tsbs_generate_data/serialize/akumuli.go:
```diff
@@ -0,0 +1,136 @@
+	buf := make([]byte, 0, 1024)
+	// Add cue
+	const HeaderLength = 8
+	buf = append(buf, "AAAAFFEE+"...)
```
can you make this a named constant? I am unclear what this magic value is.

RobAtticus commented on cmd/tsbs_generate_data/serialize/akumuli.go:
```diff
@@ -0,0 +1,136 @@
+			// Shortcut
+			s.index++
+			tmp := make([]byte, 0, 1024)
+			tmp = append(tmp, "AAAAFFEE*2\n"...)
```
same thing as above

RobAtticus commented on cmd/tsbs_generate_data/serialize/akumuli.go:
```diff
@@ -0,0 +1,136 @@
+	// Update cue
+	binary.LittleEndian.PutUint16(buf[4:6], uint16(len(buf)))
+	binary.LittleEndian.PutUint16(buf[6:HeaderLength], uint16(len(p.fieldValues)))
+	if deferPoint {
```
this is probably more go-like written as:
```go
if deferPoint {
  s.deferred = append(s.deferred, buf...)
  return nil
}
_, err = w.Write(buf)
return err
```

Lazin replied:
Done

RobAtticus commented on cmd/tsbs_generate_data/serialize/akumuli.go:
```diff
@@ -0,0 +1,136 @@
+			binary.LittleEndian.PutUint16(buf[6:HeaderLength], uint16(0))
+			binary.LittleEndian.PutUint32(buf[:4], id)
+		} else {
+			panic("Unexpected series name")
```
let's just return an error here? A long term goal I have is to reduce the amount of panics in the codebase, and given this func already returns errors, we should probably just do that

Lazin replied:
Done

RobAtticus commented:
Sorry it took so long, but merged!
